core/src/main/java/org/elasticsearch/index/query/RandomSampleQuery.java

```diff
@@ -0,0 +1,318 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.query;
+
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Objects;
+import java.util.Random;
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.BulkScorer;
+import org.apache.lucene.search.ConstantScoreScorer;
+import org.apache.lucene.search.ConstantScoreWeight;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.LeafCollector;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.Scorer;
+import org.apache.lucene.search.Weight;
+import org.apache.lucene.util.Bits;
+import org.elasticsearch.common.Randomness;
+
+/**
+ * A query that randomly matches documents with a user-provided probability.  May
+ * optionally include a seed so that matches are reproducible.
+ *
+ * This query uses two strategies depending on the size of the segment.
+ * For small segments, this uses simple random sampling per-document.  For larger
+ * segments, sampling is sped up by approximating the _gap_ between samples, then
+ * skipping forward that amount.
+ *
+ * Gap-sampling is based on the work of Jeffrey Vitter in "An Efficient Algorithm
+ * for Sequential Random Sampling" (http://www.ittc.ku.edu/~jsv/Papers/Vit87.RandomSampling.pdf),
+ * and more recently documented by Erik Erlandson
+ * (http://erikerlandson.github.io/blog/2014/09/11/faster-random-samples-with-gap-sampling/)
+ */
+public final class RandomSampleQuery extends Query {
+
+    private final double p;
+    private final Long seed;
+
+    // Above this threshold, it is probably faster to just use simple random sampling
+    private static final double PROBABILITY_THRESHOLD = 0.7;
+    private static final double EPSILON = 1e-10;
+    private static double LOG_INVERSE_P;
+
+    RandomSampleQuery(double p, Long seed) {
+        assert(p > 0.0 && p <= 1.0);
+        this.p = p;
+        this.seed = seed;
+        LOG_INVERSE_P =  Math.log(1-p);
```

Comment:
Assigning a static variable in a constructor looks buggy to me, maybe logInverseP should not be static?